src/casp/prepare_for_training.py

```diff
@@ -0,0 +1,135 @@
+"""
+Prepare data for extract by randomizing, preprocessing and staging in temporary tables
+"""
+import argparse
+import time
+
+from google.cloud import bigquery
+
+
+def execute_query(client, sql):
+    """
+    Runs the supplied query
+    """
+    start = time.time()
+
+    # print(f"RUNNING:\n {sql}")
+    query = client.query(sql)
+    results = query.result()
+
+    job = client.get_job(query.job_id)
+    gb_billed = int(0 if job.total_bytes_billed is None else job.total_bytes_billed) / (1024 * 1024 * 1024)
+    print(f"COMPLETED ({time.time() - start} seconds, {gb_billed} GBs scanned)")
+
+    return results
+
+
+def prepare_feature_summary(client, project, dataset, extract_table_prefix):
+    """
+    create feature/gene level summary -- scan of full dataset
+    """
+    sql = f"""
+        CREATE OR REPLACE TABLE `{project}.{dataset}.{extract_table_prefix}__extract_feature_summary`
+        AS
+        SELECT  f.feature_name,
+                SUM(m.raw_counts) total_raw_counts,
+                COUNT(distinct CASE WHEN m.raw_counts > 0 THEN m.cas_cell_index ELSE null END) cells_with_counts
+        FROM `{project}.{dataset}.cas_raw_count_matrix` m
+        JOIN `{project}.{dataset}.cas_feature_info` f ON (m.cas_feature_index = f.cas_feature_index)
+        GROUP BY f.feature_name
+    """
+    print("Creating Feature Summary...")
+    query = execute_query(client, sql)
+    return query
+
+
+def prepare_feature_info(client, project, dataset, extract_table_prefix, min_observed_cells):
+    """
+    create subset of features for extract with new local identifiers
+    TODO -- join to static list of ensembl ids
+    """
+    sql = f"""
+        CREATE OR REPLACE TABLE `{project}.{dataset}.{extract_table_prefix}__extract_feature_info`
+        AS
+        SELECT  DENSE_RANK() OVER (ORDER BY s.feature_name ASC) AS cas_feature_index,
+                s.feature_name as feature_name,
+        FROM	`{project}.{dataset}.{extract_table_prefix}__extract_feature_summary` s
+        WHERE s.cells_with_counts >= {min_observed_cells}
+        ORDER BY s.feature_name
+    """
+
+    print("Creating Feature Info...")
+    query = execute_query(client, sql)
+    return query
+
+
+def prepare_cell_info(client, project, dataset, extract_table_prefix, extract_bin_size):
+    """
+    Randomize cells into bins of `extract_bin_size`
+    NOTE: bins are approximately `extract_bin_size` cells
+    """
+    sql = f"""
+        CREATE OR REPLACE TABLE `{project}.{dataset}.{extract_table_prefix}__extract_cell_info`
+        AS
+        SELECT  cas_cell_index,
+                CAST(FLOOR(rand() * (select count(1) from `{project}.{dataset}.cas_cell_info`) / {extract_bin_size}) as INT) as extract_bin
+        FROM	`{project}.{dataset}.cas_cell_info` c
+        ORDER BY cas_cell_index
+    """
```

Comment:
This seems to produce a nice set of roughly equally sized bins, except for that last one...

![image](https://user-images.githubusercontent.com/10790523/190283124-a664bd45-032c-4bd4-ba21-c9cec1342954.png)

![image](https://user-images.githubusercontent.com/10790523/190283340-4ff6e241-c992-429c-829f-8b1d31820eef.png)

Reply:
I can add a comment about this, but there will be a "remainder" bin that is smaller than the rest.  There are other approaches, but a reasonable place to start for now and a we learn more about the ML requirements we can optimize.  Ie how important is it for the bin to be the exact same size, vs what you asked for, etc.